2015/presentations/sztuczki_z_metoda_super-Dominik-Socek-Dlugajczyk.md

```diff
@@ -0,0 +1,64 @@
+
+Niestety taka sztuczka ma sens tylko w testach, gdyż wtedy wiemy dokładnie jaką klasę blokujemy i jak MRO się zachowa (oraz jak struktura się zmienii). Jeśli byśmy chcieli jeszcze raz podziedziczyć po wielu klasach, to hierarchia MRO może ulec zmianie i nasz kod przestanie działać. Tak samo nie możemy zablokować kodu konkretnej klasy w hierarchii, już nie mówiąc o tym, że trzeba by przepisać wszystkie te klasy, po których nasza klasa będzie dziedziczona, aby zmienić strukturę dziedziczenia (co jest niewykonalne, gdyż nigdy nie wiemy co będzie dziedziczyło po naszej klasie).
+
+Wiedząc to wszystko bardzo łatwo będzie nam teraz zrozumieć po co w argumentach metody super trzeba podać aktualną klasę oraz aktualny obiekt. Czyli:
+
+```super(OrdersListController, self)```
+
+Pierwszym argument jest miejsce w liście MRO, drugie jest to obiekt od której lista się zaczyna. Na przykład: "self" może być typu OrdersListController, jeśli stworzyliśmy obiekt tej klasy, natomiast jeśli zrobiliśmy klasę która tylko dziedziczy po OrdersListController, to self będzie typu tej nowej klasy. Dlatego użycie super w tej formie nie ma sensu:
+```super(self.__class__, OrderListController)```
+
+Zauważyłem jakiś czas temu, że bardzo często używam super() na początku każdej metody. Chciałem zatem zrobić dekorator @superme, który by wykonywał super za mnie. Nie udało mi się to, gdyż dekoratory działają w dość prosty sposób. Dekoratory jest to funkcja, która bierze funkcję i zwraca inną funkcję. Jeśli zrobimy dekorator na metodzie, to w klasie wyciągniemy tę metodę i wystawimy nową. Jeśli spróbujemy w tej nowej funkcji użyć super, to wszystko powinno grać. Jeśli natomiast spróbujemy użyć super na metodzie, której rodzic ma metodę pod dekoratorem, to używając super nie dostaniemy naszej upragnionej metody, lecz funkcję zwróconą przez dekorator. Czyli mając coś takiego:
+
+```
+def superme(fun):
+    def wrapper(*args, **kwargs):
+        return fun(*args, **kwargs)
+    return wrapper
+
+
+class One(object):
+
+    @superme
+    def method(self):
+        pass
+
+
+class Two(One):
+
+    def method(self):
+        print(super().method)
+
+Two().method()
+```
+
+Wklasie Two.method, pod super().method nie dostaniemy One.method, tylko coś w rodzaju: "<bound method Two.wrapper of <__main__.Two object at 0x7f2116f7b048>" Mając Two.method albo One.method, możemy dostać się do klasy w której ta metoda została zdefiniowana. Mając już klasę, możemy użyć metody super(), gdyż mamy pełną listę MRO (dzięki self), oraz miejsce w hierarchii (dzięki klasie). Mając funkcję wrapper nie jesteśmy w stanie (a przynajmniej ja nie znam sposobu) dostać się do klasy, na której ten wrapper został założony.
```

Comment:
A tu Cię zaskoczę, skoro mamy functools.wraps to możemy. Wystarczy użyć trochę magii inspecta.

```
import inspect
from functools import wraps


def get_class_that_defined_method(meth):
    if inspect.ismethod(meth):
        for cls in inspect.getmro(meth.__self__.__class__):
            if cls.__dict__.get(meth.__name__) is meth:
                return cls
        meth = meth.__func__  # fallback to __qualname__ parsing
    if inspect.isfunction(meth):
        module = inspect.getmodule(meth)
        data = meth.__qualname__.split('.<locals>', 1)[0].rsplit('.', 1)[0]
        cls = getattr(module, data)
        if isinstance(cls, type):
            return cls
    return None


def superme(fun):
    @wraps(fun)
    def wrapper(*args, **kwargs):
        return fun(*args, **kwargs)
    return wrapper


class One(object):

    @superme
    def method(self):
        pass


class Two(One):

    def method(self):
        method = super(Two, self).method
        cls = get_class_that_defined_method(method)
        print(method, cls)

Two().method()
```

Tylko muszę jeszcze przemyśleć jak to wrzucić do artykułu, aby miało ręce i nogi.
Dzięki wielkie za wszystkie uwagi.